### PR TITLE
Test fix ke1xf userspace failures

### LINKED ISSUE #16225: tests/kernel/msgq/msgq_api twr_ke18f fails with assert
**Describe the bug**
assert failure

**To Reproduce**
1. mkdir build; cd build
2. cmake -DBOARD=twr_ke18f ..
3. make

**Expected behavior**
PASS

**Impact**
message queue issue

**Screenshots or console output**
```
***** Booting Zephyr OS 1.14.99 *****
ASSERTION FAIL [num_parts <= max_partitions] @ ./kernel/mem_domain.c:90
***** Kernel Panic! *****
Current thread ID = 0x2000070c
Faulting instruction address = 0xa154
```
**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain Zephyr SDK
 - Commit SHA: c8708d9

### LINKED ISSUE #16229: tests/kernel/common fails at test_atomic on twr_ke18f board
**Describe the bug**
test_atomic meets mpu failure

**To Reproduce**
1. mkdir build; cd build
2. cmake -DBOARD=twr_ke18f ..
3. make

**Expected behavior**
test PASS

**Impact**
kernel can not work well

**Screenshots or console output**
```
Running test suite common

===================================================================

starting test - test_verify_bootdelay

PASS - test_verify_bootdelay

===================================================================

starting test - test_irq_offload

PASS - test_irq_offload

===================================================================

starting test - test_byteorder_memcpy_swap

PASS - test_byteorder_memcpy_swap

===================================================================

starting test - test_byteorder_mem_swap

PASS - test_byteorder_mem_swap

===================================================================

starting test - test_atomic

ASSERTION FAIL [0] @ ./arch/arm/core/cortex_m/mpu/nxp_mpu.c:563

	Configuring 3 dynamic MPU regions failed



[00:00:00.560,000] [1;31m<err> mpu: Failed to allocate new MPU region 8
[0m

***** Kernel Panic! *****

Current thread ID = 0x2000077c

Faulting instruction address = 0xda4c

```
**Environment (please complete the following information):**
 - OS: ( Linux )
 - Toolchain (Zephyr SDK, ...)
 - Commit SHA: c8708d9

### LINKED ISSUE #18284: tests/kernel/fp_sharing/float_disable and tests/kernel/mutex/mutex_api and tests/kernel/sleep  fails on twr_ke18f
**Describe the bug**
meet same memory region issue with user mode enalbed

**To Reproduce**
1. mkdir build; cd build
2. cmake -DBOARD=twr_ke18f ..
3. make

**Expected behavior**
PASS

**Impact**
float point usage

**Screenshots or console output**
```
***** Booting Zephyr OS version 2.0.0-rc1 *****

ASSERTION FAIL [num_parts <= max_partitions] @ ZEPHYR_BASE/kernel/mem_domain.c:90

	

FATAL: r0/a1:  0x00000004  r1/a2:  0x0000005a  r2/a3:  0x00800000

FATAL: r3/a4:  0x00000000 r12/ip:  0x00000008 r14/lr:  0x00006ea3

FATAL:  xpsr:  0x61000000

FATAL: s[0]:  0x00000000  s[1]:  0x00000000  s[2]:  0x00000000  s[3]:  0x00000000



FATAL: s[4]:  0x00000000  s[5]:  0x00000000  s[6]:  0x00000000  s[7]:  0x00000000



FATAL: s[8]:  0x00000000  s[9]:  0x00000000  s[10]:  0x00000000  s[11]:  0x00000000



FATAL: s[12]:  0x00000000  s[13]:  0x00000000  s[14]:  0x00000000  s[15]:  0x00000000



FATAL: fpscr:  0x00000000



FATAL: Faulting instruction address (r15/pc): 0x00009a96

FATAL: >>> ZEPHYR FATAL ERROR 4: Kernel panic

FATAL: Current thread: 0x20000454 (main)

FATAL: Halting system

```
**Environment (please complete the following information):**
 - OS: (e.g. Linux, MacOS, Windows)
 - Toolchain (e.g Zephyr SDK, ...)
 - R2.0.0-rc1

### PR DESCRIPTION
**Pull request by ioannisg**
Add an option to exlude HW Stack protection from user mode tests. This allows us to run the user mode test suite for platforms with limited number of MPU regions, such as the twr_ke18f board.
Fixes #18284 
Fixes #16229
Fixes #16225

### CODE REVIEW

**Comment by henrikbrixandersen** [comment]
@ioannisg Thank you for working on this. I have verified the fix on my TWR-KE18F board for a few select test cases. I will start a sanitycheck and report back.

**Comment by henrikbrixandersen** [comment]
> I would like to understand why ke1xf defines so many hard-coded MPU regions (five) in nxp_mpu_regions.c.

@andrewboie Please see my comment at https://github.com/zephyrproject-rtos/zephyr/issues/16229#issuecomment-495947933

**Comment by henrikbrixandersen** [comment]
> @ioannisg Thank you for working on this. I have verified the fix on my TWR-KE18F board for a few select test cases. I will start a sanitycheck and report back.

The following tests still fail due to too few free MPU regions:
- tests/arch/arm/arm_ramfunc/arch.ramfunc
- tests/kernel/mem_protect/mem_protect/kernel.memory_protection
- tests/kernel/mem_protect/obj_validation/kernel.memory_protection.obj_validation
- tests/kernel/mem_protect/syscalls/kernel.memory_protection.syscalls
- tests/lib/mem_alloc/libraries.libc.newlib

The following tests fail with NXP MPU bus faults:
- samples/userspace/shared_mem/sample.kernel.memory_protection.shared_mem
- tests/kernel/fatal/kernel.common.stack_protection
- tests/kernel/fatal/kernel.common.stack_protection_arm_fp_sharing

**Review by agansari — APPROVED** [review]

**Comment by ioannisg** [comment]
@henrikbrixandersen could you please check with current version of the PR - how many additional tests failures are resolved?

**Comment by henrikbrixandersen** [comment]
> @henrikbrixandersen could you please check with current version of the PR - how many additional tests failures are resolved?

The following test still fails due to too few free MPU regions:
- tests/lib/mem_alloc/libraries.libc.newlib

The following tests fail with NXP MPU bus faults:
- tests/kernel/fatal/kernel.common.stack_protection
- samples/userspace/shared_mem/sample.kernel.memory_protection.shared_mem
- tests/kernel/mem_protect/mem_protect/kernel.memory_protection

**Review by galak — CHANGES_REQUESTED** [review]
Is this because of number of MPU regions?

See comment about where we enable this.

File: soc/arm/nxp_kinetis/ke1xf/Kconfig.defconfig.series [hunk 1]
```diff
@@ -108,6 +108,9 @@ config ADC_MCUX_ADC12
 
 endif # ADC
 
+config TEST_USERSPACE_WITHOUT_HW_STACK_PROTECTION
```
[the remark this answers is not in the document]
Comment (ioannisg) [reply soc/arm/nxp_kinetis/ke1xf/Kconfig.defconfig.series]: @galak to clarify: you prefer adding `default y if SOC_SERIES_KINETIS_KE1XF` in the original definition of `TEST_USERSPACE_WITHOUT_HW_STACK_PROTECTION `in subsys/testsuite/Kconfig?

**Comment by ioannisg** [comment]
> 
> 
> > @henrikbrixandersen could you please check with current version of the PR - how many additional tests failures are resolved?
> 
> The following test still fails due to too few free MPU regions:
> 
>     * tests/lib/mem_alloc/libraries.libc.newlib
> 
This one is rather simple - we can exclude the platform for this test.

File: soc/arm/nxp_kinetis/ke1xf/Kconfig.defconfig.series (hunk 1 above)
[the remark this answers is not in the document]
Comment (ioannisg) [reply soc/arm/nxp_kinetis/ke1xf/Kconfig.defconfig.series]: that's fine with me, will apply this proposal.

**Review by MaureenHelm — COMMENTED** [review]
> I would like to understand why ke1xf defines so many hard-coded MPU regions (five) in nxp_mpu_regions.c.

k64f also uses 5 regions, but it has 12 regions available in hardware whereas ke18f has only 8.

> Does the NXP MPU have no concept of a background System map?

No

> Why do we need an address-space sized region for "DEBUGGER_0"?

To access memory with a debugger.

> Why do we need a region for "RAM_U_0", given that supervisor mode ought to already be able to read/write RAM?
Can we disable "BACKGROUND_0" and "BACKGROUND_1" if CONFIG_HW_STACK_PROTECTION is disabled?

Yes, this may be possible. We should wait to do this until after the 2.0 release because it's really easy to break things when messing with MPU regions. I propose we merge this PR now and open an enhancement issue to optimize the number of boot MPU regions.

> My feeling is that the boot MPU region config for a lot of boards were set up long before userspace was even implemented and need to be revisited.

File: soc/arm/nxp_kinetis/ke1xf/Kconfig.defconfig.series (hunk 1 above)
[the remark this answers is not in the document]
Comment (ioannisg) [reply soc/arm/nxp_kinetis/ke1xf/Kconfig.defconfig.series]: done

**Comment by ioannisg** [comment]
> 
> 
> > @henrikbrixandersen could you please check with current version of the PR - how many additional tests failures are resolved?
> 
> The following test still fails due to too few free MPU regions:
> 
>     * tests/lib/mem_alloc/libraries.libc.newlib

This test uses one extra region for Heap, which we do not have, even if we disable stack protection. So I will exclude the platform for this test.
>

**Comment by ioannisg** [comment]
>The following tests fail with NXP MPU bus faults:
>
>    tests/kernel/fatal/kernel.common.stack_protection
>    samples/userspace/shared_mem/sample.kernel.memory_protection.shared_mem
>    tests/kernel/mem_protect/mem_protect/kernel.memory_protection

These fail because they expect the HW stack protection to function. But we disable it. 
So we need to exclude them from running on this board.

**Comment by ioannisg** [comment]
@henrikbrixandersen @MaureenHelm @galak could you take a look one more time, here?

In brief, I applied as little filtering as it was possible, for this platform. Still there are around 30 user mode tests executed for this board, so we keep coverage at decent level.

**Review by MaureenHelm — APPROVED** [review]

**Comment by henrikbrixandersen** [comment]
> @henrikbrixandersen @MaureenHelm @galak could you take a look one more time, here?
> 
> In brief, I applied as little filtering as it was possible, for this platform. Still there are around 30 user mode tests executed for this board, so we keep coverage at decent level.

I am rerunning sanitycheck against the twr_ke18f board now.

**Review by henrikbrixandersen — APPROVED** [review]

**Review by galak — APPROVED** [review]

**Pull request merged**